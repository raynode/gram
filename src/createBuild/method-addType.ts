import {
  BuildModeArgsGenerator,
  BuildModeGenerator,
  CreateableTypes,
  Fields,
} from './types'

import { isBuildModeGenerator } from './guards'

export type Definable<BuildMode, Data> =
  | Data
  | BuildModeGenerator<BuildMode, Data>
export type DefinableFields<BuildMode> = Definable<BuildMode, Fields>

export type AddScalarArgsType = [string, 'scalar', never]
export type AddInterfaceArgsType<BuildMode> = [
  string,
  'interface',
  { fields: DefinableFields<BuildMode> }
]
export type AddInputArgsType<BuildMode> = [
  string,
  'input',
  { fields: DefinableFields<BuildMode> }
]
export type AddEnumArgsType<BuildMode> = [
  string,
  'enum',
  { values: Definable<BuildMode, string[]> }
]
export type AddObjectArgsType<BuildMode> = [
  string,
  'type',
  {
    fields: DefinableFields<BuildMode>;
    interface?: Definable<BuildMode, string>;
  }
]

export type AddArgs<BuildMode> =
  | AddScalarArgsType
  | AddInterfaceArgsType<BuildMode>
  | AddInputArgsType<BuildMode>
  | AddEnumArgsType<BuildMode>
  | AddObjectArgsType<BuildMode>

export type AddScalarType = (
  typeName: string,
  type: 'scalar',
  config?: {},
) => void
export type AddInterfaceType<BuildMode> = (
  typeName: string,
  type: 'interface',
  config: {
    fields: DefinableFields<BuildMode>;
  },
) => void
export type AddInputType<BuildMode> = (
  typeName: string,
  type: 'input',
  config: {
    fields: DefinableFields<BuildMode>;
  },
) => void
export type AddEnumType<BuildMode> = (
  typeName: string,
  type: 'enum',
  config: {
    values: Definable<BuildMode, string[]>;
  },
) => void
export type AddObjectType<BuildMode> = (
  typeName: string,
  type: 'type',
  config: {
    fields: DefinableFields<BuildMode>;
    interface?: Definable<BuildMode, string>;
  },
) => void

export type AddType<BuildMode> = AddScalarType &
  AddInterfaceType<BuildMode> &
  AddInputType<BuildMode> &
  AddEnumType<BuildMode> &
  AddObjectType<BuildMode>

export type AddArgsGenerator<BuildMode> = BuildModeArgsGenerator<
  BuildMode,
  AddArgs<BuildMode>
>

export const createAddType = <BuildMode>(
  buildMode: BuildMode,
  addScalarType: AddScalarType,
  addInterfaceType: AddInterfaceType<BuildMode>,
  addInputType: AddInputType<BuildMode>,
  addEnumType: AddEnumType<BuildMode>,
  addObjectType: AddObjectType<BuildMode>,
): AddType<BuildMode> & AddArgsGenerator<BuildMode> => (
  buildModeBenerator,
  ...args
) => {
  const typeInformation = isBuildModeGenerator<BuildMode, AddArgs<BuildMode>>(
    buildModeBenerator,
  )
    ? buildModeBenerator(buildMode)
    : ([buildModeBenerator, ...args] as AddArgs<BuildMode>)
  switch (typeInformation[1]) {
    case 'scalar':
      return addScalarType(...typeInformation)
    case 'interface':
      return addInterfaceType(...typeInformation)
    case 'input':
      return addInputType(...typeInformation)
    case 'enum':
      return addEnumType(...typeInformation)
    case 'type':
      return addObjectType(...typeInformation)
  }
}
