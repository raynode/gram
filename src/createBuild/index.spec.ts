import { graphql, GraphQLNonNull, GraphQLString, printSchema } from 'graphql'
import { makeExecutableSchema } from 'graphql-tools'
import { createBuild } from '.'
import { AddObjectArgsType, AddObjectType } from './method-addType'

const initialPets = [
  {
    name: 'Fluffy',
    kind: 'dog',
    age: 5,
    owner: 'Mark',
  },
]

const data = {
  pets: initialPets,
}

describe('createBuild', () => {
  it('should create a base object with methods to add everything for gql', () => {
    const build = createBuild()
    expect(build).toMatchSnapshot()
  })

  it('should be able to create a simple graphql interface', async () => {
    const build = createBuild()
    build.addQuery('test', GraphQLString, () => 'TEST')
    const { typeDefs, resolvers } = build.toTypeDefs()
    const schema = makeExecutableSchema({
      typeDefs,
      resolvers,
    })
    const result = await graphql({ schema, source: '{ test }' })
    expect(result).toMatchSnapshot()
  })

  it('should generate the schema directly', async () => {
    const build = createBuild()
    build.addQuery('test', GraphQLString, () => 'TEST')
    const result = await graphql({
      schema: build.toSchema(),
      source: '{ test }',
    })
    expect(result).toMatchSnapshot()
  })

  it('should accept a buildMode function', async () => {
    const build = createBuild()
    build.addQuery('test', buildMode => GraphQLString, {
      resolver: () => () => 'TEST',
    })
    const { typeDefs, resolvers } = build.toTypeDefs()
    const schema = makeExecutableSchema({
      typeDefs,
      resolvers,
    })
    const result = await graphql({ schema, source: '{ test }' })
    expect(result).toMatchSnapshot()
  })

  it('should be able to accept a new scalar type', async () => {
    const build = createBuild()
    build.addType('MyScalar', 'scalar')
    build.addQuery('myScalar', 'MyScalar', () => 'TEST')
    const result = await graphql({
      schema: build.toSchema(),
      source: '{ myScalar }',
    })
    expect(result).toMatchSnapshot()
  })

  it('should be able to accept a new object type', async () => {
    const build = createBuild()
    build.addType('Pet', 'type', {
      fields: {
        name: GraphQLString,
        age: 'Int!',
      },
    })
    build.addQuery('myPets', '[Pet!]!', () => data.pets)
    const result = await graphql({
      schema: build.toSchema(),
      source: '{ myPets { name age } }',
    })
    expect(result).toMatchSnapshot()
  })

  it('should have types that are different based on buildMode', async () => {
    type BuildMode = 'admin' | 'user'
    const adminBuild = createBuild<BuildMode>('admin')
    const userBuild = createBuild<BuildMode>('user')

    const petFields = (buildMode: BuildMode) => ({
      name: GraphQLString,
      age: 'Int!',
      ...(buildMode === 'admin' && {
        owner: GraphQLNonNull(GraphQLString),
      }),
    })

    adminBuild.addType('Pet', 'type', { fields: petFields })
    userBuild.addType('Pet', 'type', { fields: petFields })
    adminBuild.addQuery('myPets', '[Pet!]!', () => data.pets)
    userBuild.addQuery('myPets', '[Pet!]!', () => data.pets)
    expect(adminBuild.toTypeDefs().typeDefs).toMatchSnapshot()
    expect(userBuild.toTypeDefs().typeDefs).toMatchSnapshot()
  })

  it('should accept build mode generators on type level', async () => {
    type BuildMode = 'admin' | 'user'
    const adminBuild = createBuild<BuildMode>('admin')
    const userBuild = createBuild<BuildMode>('user')

    const Pet = (buildMode: BuildMode): AddObjectArgsType<BuildMode> => [
      'Pet',
      'type',
      {
        fields: {
          name: GraphQLString,
          age: 'Int!',
          ...(buildMode === 'admin' && {
            owner: GraphQLNonNull(GraphQLString),
          }),
        },
      },
    ]

    adminBuild.addType(Pet)
    userBuild.addType(Pet)
    adminBuild.addQuery('myPets', '[Pet!]!', () => data.pets)
    userBuild.addQuery('myPets', '[Pet!]!', () => data.pets)
    expect(adminBuild.toTypeDefs().typeDefs).toMatchSnapshot()
    expect(userBuild.toTypeDefs().typeDefs).toMatchSnapshot()
  })
})
