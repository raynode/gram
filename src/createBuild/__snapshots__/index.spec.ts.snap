// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBuild should accept a buildMode function 1`] = `
Object {
  "data": Object {
    "test": "TEST",
  },
}
`;

exports[`createBuild should accept build mode generators on type level 1`] = `
"
    
    
    
  type Pet {
    name: String
age: Int!
owner: String!
  }

    
  type Query {
    myPets: [Pet!]!
  }

    
    
    
  "
`;

exports[`createBuild should accept build mode generators on type level 2`] = `
"
    
    
    
  type Pet {
    name: String
age: Int!
  }

    
  type Query {
    myPets: [Pet!]!
  }

    
    
    
  "
`;

exports[`createBuild should be able to accept a new object type 1`] = `
Object {
  "data": Object {
    "myPets": Array [
      Object {
        "age": 5,
        "name": "Fluffy",
      },
    ],
  },
}
`;

exports[`createBuild should be able to accept a new scalar type 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Cannot query field "test" on type "Query".],
  ],
}
`;

exports[`createBuild should be able to create a simple graphql interface 1`] = `
Object {
  "data": Object {
    "test": "TEST",
  },
}
`;

exports[`createBuild should create a base object with methods to add everything for gql 1`] = `
Object {
  "addMutation": [Function],
  "addQuery": [Function],
  "addSubscription": [Function],
  "addType": [Function],
  "buildMode": undefined,
  "builder": "Build",
  "toSchema": [Function],
  "toTypeDefs": [Function],
}
`;

exports[`createBuild should generate the schema directly 1`] = `
Object {
  "data": Object {
    "test": "TEST",
  },
}
`;

exports[`createBuild should have types that are different based on buildMode 1`] = `
"
    
    
    
  type Pet {
    name: String
age: Int!
owner: String!
  }

    
  type Query {
    myPets: [Pet!]!
  }

    
    
    
  "
`;

exports[`createBuild should have types that are different based on buildMode 2`] = `
"
    
    
    
  type Pet {
    name: String
age: Int!
  }

    
  type Query {
    myPets: [Pet!]!
  }

    
    
    
  "
`;
